src/LOAuth.ts

```diff
@@ -111,16 +131,28 @@ class LOAuth {
     state = JSON.stringify(state);
 
     /**
-     * Skip over the encoding when there is a 'state' query parameter.  This is the
-     * scenario when we are loading from the redirected auth endpoint and the state
-     * is already encoded in the URL parameter.  The problem is that encoding/decoding
-     * is being handled on multiple layers so that this client loads after the auth
-     * redirect, client-oauth will parse the state from the query param as unencoded,
-     * stringified JSON and when we pass an encoded version in, there is a comparison
-     * failure which results in a second redirect.
+     * When the auth flow is initiated there will be no 'state' query parameter.
+     * Because it is stringified JSON it needs to be encoded before passing it to the
+     * auth provider. If the user has to go through a sign in page to be
+     * authenticated, the redirect seems to contain something close to an encoded
+     * JSON string so that 'queryParameters.get' will return stringified JSON without
+     * an encoding.  In that case we don't want to encode the state value because it
+     * will result in a failed diff when checking the value we pass to the value from
+     * the URL.
+     * If the user is already authenticated and we are redirected immediately from an
+     * auth server (no login page seen), the redirect contains the state value but it
+     * is doubly encoded.  In that case we should encode the state value so that it
+     * will match when compared to the query string
      */
     const queryParameters = new URLSearchParams(window.location.search);
-    if (!queryParameters.get('state')) {
+    const currState = queryParameters.get('state');
+
+    // Encode state if there is no query paramater or if it was doubly encoded
+    if (
+      !currState ||
+      (!isStringifiedJSON(currState) &&
+        isStringifiedJSON(decodeURIComponent(currState)))
+    ) {
       state = encodeURIComponent(state);
```

Comment:
I think that his coding might also be related to the double encoding of state. Ultimately, the state returned from this method is passed to `js-client-oauth2` library which is also encoding the state: https://github.com/mulesoft/js-client-oauth2/blob/f76a5e0dad58cf60086eb213ae2e547fd2debf95/src/client-oauth2.js#L163

Reply:
I've remove the 'query param' part of the change and added some examples to the comments to be more explicit about why the change is needed.  It is possible that there is some bug in Cognito with how it hands off to a third party auth provider like Google - that seems to be where the problems come into play.
Another alternative would be doing a base64 encoding and then not needing to worry about any of the encodeURI component pieces altogether.

Reply:
Your new version is definitely simpler and more focused on the new feature you are adding. I like that. 

@markdlv I'm still a little concerned about the double encoding of state. Specially, the (existing) used of `encodeURIComponent` here seems like it causes trouble because `js-client-oauth2` also encodes.

I'll approve without trying to fix the encoding issue, but I think it's worth fixing

Reply:
I agree.  I like the idea of base64 encoding state.  Would you have time to go for that in this PR, Dan?  We do that elsewhere (I can show you offline) and it works well.  I'm also interested in dropping `js-client-oauth2`, but I can tackle that in a separate effort/PR.

Reply:
Yeah I think we should just go with base64 - I can update this PR to do that instead of all the URI encoding